RFC: nixos/borgbackup: run as DynamicUser

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

There should not be the need to run our favorite backup tool as root.
Why don't we execute it as a DynamicUser and give it permissions to
bypass discrectionary access control.

```
       CAP_DAC_READ_SEARCH
              * Bypass file read permission checks and directory read
                and execute permission checks;
              * invoke open_by_handle_at(2);
              * use the linkat(2) AT_EMPTY_PATH flag to create a link to
                a file referred to by a file descriptor.
```
*capabitilies(7)*

This would still be a reduced surface area, than what we have today.

Instead of embedding the pre and post hook into the backup script we
move them to ExecStartPre and ExecStopPost and grant them full
privileges, so they can still achieve tasks like create and delete
snapshots. They are short running scripts and have less surface area
than the whole of borgbackup.

This doesn't come without breaking changes. The `archiveName` could
previously be passed in the `preHook` option, but that will stop working
as soon as we move the hooks into their own scripts. We could alleviate
this problem by setting up an explicit `archiveName` option, that is
being passed into the main backup script.

There is currently also the issue, that the `local` nixos tests job will
not run, because borgbackup cannot write to the local backup location.
This could be fixed by adding the location to `readWritePaths`, but
first we would have to determine if it was actually a local path, and
not a remote one.

Also some users might already have configured borgbackup to have more
constrained user and group settings, so I could imagine this could
extend the permission of those jobs. But I think this could be a
more secure default for most people.

The whole idea apparently isn't new, it came up over at the borg-
backup upstream issue tracker back in 2017.

https://github.com/borgbackup/borg/issues/2125

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mweinelt committed:
nixos/borgbackup: optionally run as DynamicUser

When setting user to `null` the systemd unit uses a DynamicUser setup
with the CAP_DAC_READ_SEARCH capability, thereby being able to bypass
filesystem read and search permission checks.

To still be able to achieve privileged pre and post tasks, they are
moved into ExecStartPre and ExecStopPost respectively.

This doesn't come without breaking changes. The custom `exitStatus`
variable is being dropped in favor of the inherited `EXIT_STATUS` and
`EXIT_CODE`, that are provided by systemd itself. The exit trap itself
seems to have only been created for cleanup tasks, which are moved to
ExecStopPost, and the `exitStatus` variable, so it has been dropped
entirely.

Mic92 commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -78,8 +76,13 @@ let
-        User = cfg.user;
-        Group = cfg.group;
+        ExecStartPre = "+${pkgs.writeShellScript "borgbackup-pre-${name}" cfg.preHook}";
+        ExecStopPost = "+${pkgs.writeShellScript "borgbackup-post-${name}" cfg.postHook}";
```
post hooks are also used for monitoring. Is there a way to still get `exitStatus` in this case? I currently do this: https://github.com/Mic92/dotfiles/blob/e22911ae4d929bc91316a2c4a4112ad175b5bc77/nixos/matchbox/modules/borgbackup.nix#L34

mweinelt replied:
>  Is ExecStopPost executed in both the error and success case?

Yes, it is a cleanup job basically.

> post hooks are also used for monitoring. Is there a way to still get exitStatus in this case? 

I suspect we could move that use case into `ExecStartPost`. This would obviously entail that we change the nomenclature on the hooks a bit.

mweinelt replied:
Looks like `EXIT_STATUS` is what you are looking for.

```
client # [   36.370212] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: extraArgs=
client # [   36.373147] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: LANG=en_US.UTF-8
client # [   36.375511] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: SERVICE_RESULT=success
client # [   36.378141] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: extraPruneArgs=
client # [   36.380297] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: INVOCATION_ID=2cc515adada14ffc99b45965e1c43319
client # [   36.381879] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: EXIT_STATUS=0
client # [   36.383178] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: BORG_REPO=borg@server:.
client # [   36.384531] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: USER=borgbackup-job-remote
client # [   36.386276] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: TZDIR=/nix/store/yijklpbih24arzw7jcc3ykwbndd8gbgf-tzdata-2020f/share/zoneinfo
client # [   36.388193] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: PWD=/
client # [   36.389428] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: JOURNAL_STREAM=8:13995
client # [   36.390853] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: BORG_BASE_DIR=/var/lib/borgbackup
client # [   36.392377] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: extraInitArgs=
client # [   36.399521] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: extraCreateArgs=
client # [   36.401899] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: STATE_DIRECTORY=/var/lib/borgbackup
client # [   36.404020] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: SHLVL=1
client # [   36.405827] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: EXIT_CODE=exited
client # [   36.407573] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: LOCALE_ARCHIVE=/nix/store/hvjscb69218cw0r60pg7szxv8d5dk6lz-glibc-locales-2.32-40/lib/locale/locale-archive
client # [   36.410466] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: LOGNAME=borgbackup-job-remote
client # [   36.413138] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: BORG_RSH=ssh -oStrictHostKeyChecking=no -i /root/id_ed25519
client # [   36.415573] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: PATH=/nix/store/01xwxvlyavk7n8gpmmjbryrjflgs4cl9-borgbackup-1.1.16/bin:/nix/store/6izw7xiy7f7lib7dsqn2ip476kmpkzm5-openssh-8.5p1/bin:/nix/store/lr96h3dlny8aiba9p3rmxcxfda0ijj08-coreutils-8.32/bin:/nix/store/6kjrp5p21icxr38zws6yzbdsbf8z66b6-findutils-4.7.0/bin:/nix/store/3v5i98i92j0f3lbb7d58kvf8nxnhw7s7-gnugrep-3.6/bin:/nix/store/4nf4ih03fcq7gk08spjzxvwph1vyx1kr-gnused-4.8/bin:/nix/store/ixpwj1cxl4rp2qbyn0s2h4k5zj731q7c-systemd-247.6/bin:/nix/store/01xwxvlyavk7n8gpmmjbryrjflgs4cl9-borgbackup-1.1.16/sbin:/nix/store/6izw7xiy7f7lib7dsqn2ip476kmpkzm5-openssh-8.5p1/sbin:/nix/store/lr96h3dlny8aiba9p3rmxcxfda0ijj08-coreutils-8.32/sbin:/nix/store/6kjrp5p21icxr38zws6yzbdsbf8z66b6-findutils-4.7.0/sbin:/nix/store/3v5i98i92j0f3lbb7d58kvf8nxnhw7s7-gnugrep-3.6/sbin:/nix/store/4nf4ih03fcq7gk08spjzxvwph1vyx1kr-gnused-4.8/sbin:/nix/store/ixpwj1cxl4rp2qbyn0s2h4k5zj731q7c-systemd-247.6/sbin
client # [   36.428538] 9ld91vi1kc745zqc0g3fndc1a9x7db2z-borgbackup-post-remote[903]: _=/nix/store/lr96h3dlny8aiba9p3rmxcxfda0ijj08-coreutils-8.32/bin/env
```

dotlambda replied:
Documented here: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#%24EXIT_CODE

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -92,6 +95,7 @@ let
         PrivateTmp = cfg.privateTmp;
       };
       environment = {
+        HOME = "/var/lib/borgbackup";
```
```suggestion
        BORG_BASE_DIR = "/var/lib/borgbackup";
```

mweinelt replied:
Applied. Thanks!

dotlambda reviewed (commented):
If we do this, shouldn't we restrict read access to only those paths we want to back up? This is not an improvement for those who already don't run it as root. But we could just set `DynamicUser` only if `user` is not set.

dotlambda commented:
If we're already touching this service, let's split repos and jobs into two different files.

mweinelt commented:
> If we do this, shouldn't we restrict read access to only those paths we want to back up?


With `cap_dac_read_search` we bypass all permissions checks and are able to read everything. Without it and with only `ReadOnlyPaths` we would still rely on file and directory permissions. So I'm not exactly sure how to accomplish this.

mweinelt commented:
> But we could just set DynamicUser only if user is not set.

The logic could be, that if you set `cfg.user` or `cfg.group` we don't set the cap for you, because you clearly want to constrain things manually, but instead set `ReadOnlyPaths` for `cfg.paths`.

If you didn't set any of that we run as `DynamicUser` and you can read whatever path.

mweinelt commented:
> If we're already touching this service, let's split repos and jobs into two different files.

I consider this orthogonal to the idea of this pull request, but I can certainly do that, when everything else is fleshed out.

dotlambda commented:
> > If we do this, shouldn't we restrict read access to only those paths we want to back up?
> 
> With `cap_dac_read_search` we bypass all permissions checks and are able to read everything. Without it and with only `ReadOnlyPaths` we would still rely on file and directory permissions. So I'm not exactly sure how to accomplish this.

I think the only way to achieve this is https://www.freedesktop.org/software/systemd/man/systemd.exec.html#TemporaryFileSystem=. But I guess it's a little overkill.
People who want it can use the `confinement` option.

> > But we could just set DynamicUser only if user is not set.
> 
> The logic could be, that if you set `cfg.user` or `cfg.group` we don't set the cap for you, because you clearly want to constrain things manually, but instead set `ReadOnlyPaths` for `cfg.paths`.
> 
> If you didn't set any of that we run as `DynamicUser` and you can read whatever path.

That's what I had in mind.

dotlambda commented:
> > If we're already touching this service, let's split repos and jobs into two different files.
> 
> I consider this orthogonal to the idea of this pull request, but I can certainly do that, when everything else is fleshed out.

I can just do it in a different pull request, once this one is merged.

mweinelt committed:
nixos/tests: update to work with DynamicUser change

The local job wants to write to /root/back:up, so it is crucial that it
runs as root, make that explicit.

The `archiveName` option cannot be modified in `preHook` any longer,
since it is not inserted into the backup script any longer. Instead we
parse the archive name dynamically from `borg list` now.

Converted the local jobset to f'-strings, to make them easier to parse.

mweinelt commented:
Updated. The tests are working now, the local test used to override `archiveName` within `preHook`, that's not possibly anymore.

mweinelt commented:
```
  userHome = if (isNull cfg.user) then "/var/lib/borgbackup" else config.users.users.${cfg.user}.home;
```

How can I rewrite this to work, when `${cfg.user}` is null?

> while evaluating the attribute 'users.users."${(cfg).user}".home' at undefined position:
> value is null while a string was expected

mweinelt commented:
Another question would be: DynamicUser or fixed user? You probably can't give a DynamicUser write permissions to anywhere on the local disk. Is this a usecase where someone should specify explicit user/group? Or should we just allocate the borgbackup user/group?

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -92,6 +90,7 @@ let
         PrivateTmp = cfg.privateTmp;
       };
       environment = {
+        BORG_BASE_DIR = userHome;
```
This should be added to `mkBorgWrapper`, though I really don't like that duplication. Neither do I like `mkBorgWrapper` at all.

mweinelt replied:
Yeah, then we have to duplicate `userHome` logic.

dotlambda replied:
Actually, we should only set `environment.BORG_BASE_DIR` if using a dynamic user.

mweinelt replied:
Done.

dotlambda commented:
> How can I rewrite this to work, when `${cfg.user}` is null?

Use `lib.getAttrByPath`.

> Another question would be: DynamicUser or fixed user? You probably can't give a DynamicUser write permissions to just about anywhere on the local disk. Is this a usecase where someone should specify explicit user/group? Or should we just allocate the borgbackup user/group?

What's the use case for local backups anyway? I can only think of a udisks rule that starts a backup when an external drive is plugged in.
If we want to continue supporting that use case, we should just use the `borg` user that we also use for repos. I wonder why we didn't allocate a static uid for that user.

dotlambda commented:
@mweinelt I guess you don't plan on getting this into 21.05?

mweinelt commented:
Yes, too many other things I'm working on, and I think such a change needs proper testing, as it could easily break expectations.

hyperfekt commented:
I just started implementing the same thing for restic, with `cap_dac_read_search` and `TemporaryFilesystem`. For local backups either setfacl or bindfs might be an option, but for now I'll see if I can just detect and exclude those cases.
What is this `confinement` mentioned by @dotlambda? Is that borg-specific?

hyperfekt commented:
After further research, combining `TemporaryFilesystem` (or the confinement option, which uses it as well) with `DynamicUser` is currently blocked on systemd/systemd#18999.

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -70,16 +59,25 @@ let
       script = mkBackupScript cfg;
-      serviceConfig = {
+      serviceConfig = optionalAttrs (!isNull cfg.user) {
         User = cfg.user;
```
This should be `"borg"` if using a dynamic user so systemd assigns a nice name to that user.

dotlambda replied:
Actually, that might conflict with the borg user for repos, so something like `"borg-dynamic"` might be better?

mweinelt replied:
That is the section for when a user is given. I've added the `borgbackup` user to the DynamicUser section below.

mweinelt commented:
> If we do this, shouldn't we restrict read access to only those paths we want to back up? This is not an improvement for those who already don't run it as root. But we could just set `DynamicUser` only if `user` is not set.

I think we covered this point in that this is an improvement for everyone currently running with root currently. Everyone who sets `user` themselves is out of scope.

> Documented here: freedesktop.org/software/systemd/man/systemd.exec.html#%24EXIT_CODE

Breaking change documented.

> > Another question would be: DynamicUser or fixed user? You probably can't give a DynamicUser write permissions to just about anywhere on the local disk. Is this a usecase where someone should specify explicit user/group? Or should we just allocate the borgbackup user/group?
> 
> What's the use case for local backups anyway? I can only think of a udisks rule that starts a backup when an external drive is plugged in.
> If we want to continue supporting that use case, we should just use the `borg` user that we also use for repos. I wonder why we didn't allocate a static uid for that user.

Previously local backups just worked when relying on the root user. Now you have to specify a specific user to satisfy permissions, because the DynamicUser by itself won't be able to write just anywhere.

I'm not sure if we should fix the default behaviour to restore that functionality, or if we should just document this as a breaking change.

---

I'm sure I lost track of things to do, but the scope is already quite interesting.

Might warrant a small test case.

mweinelt commented:
> > How can I rewrite this to work, when `${cfg.user}` is null?
> 
> Use `lib.getAttrByPath`.

That part is still unresolved. Looking into it now.

winterqt commented:
Still relevant -- not stale.

dotlambda commented on nixos/modules/services/backup/borgbackup.nix:
```diff
@@ -510,7 +511,7 @@ in {
-              The latter is available as <literal>$exitStatus</literal>.
+              The latter is available as <literal>$EXIT_CODE</literal>.
             '';
             default = "";
```
```suggestion
            default = null;
```
